platform/pulp/platform/models/content.py

```diff
@@ -0,0 +1,94 @@
+"""
+Content related Django models.
+"""
+import hashlib
+
+from django.contrib.contenttypes import fields
+from django.db import models
+
+
+from pulp.platform.models import Model, MasterModel, Notes
+from pulp.platform.models.storage import StoragePath
+
+
+class Content(MasterModel):
+    """
+    A piece of managed content.
+
+    :cvar natural_key_fields: Tuple of natural fields.  Must be: models.Field.
+    :type natural_key_fields: tuple
+
+    Fields:
+
+    :cvar type: The content type.
+    :type type: models.TextField
+
+    Relations:
+
+    :cvar notes: Arbitrary information stored with the content.
+    :type notes: fields.GenericRelation
+    """
+    natural_key_fields = ()
+
+    type = models.TextField(blank=False, default=None)
+
+    notes = fields.GenericRelation(Notes)
+
+    def natural_key(self):
+        return (getattr(self, f.name) for f in self.natural_key_fields)
+
+    def natural_key_digest(self):
+        h = hashlib.sha256()
+        for name in sorted(f.name for f in self.natural_key_fields):
+            h.update(name)
+            value = getattr(self, name)
+            if not isinstance(value, basestring):
+                h.update(str(value))
```

Comment:
I don't believe this is Python 3 compatible. Django returns Unicode codepoints and you cannot feed those to hashlib:
```
(pulp3) [vagrant@dev platform]$ python3 manage.py shell
Python 3.5.1 (default, Aug  9 2016, 15:35:51) 
[GCC 6.1.1 20160621 (Red Hat 6.1.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from pulp.platform.models import Repository
>>> repo = Repository(name='test')
>>> repo.save()
>>> repos = Repository.objects.filter()
>>> repos[0].name
'test'
>>> import hashlib
>>> h = hashlib.sha256()
>>> h.update(repos[0].name)
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/pulp3/lib/python3.5/site-packages/Django-1.8.14-py3.5.egg/django/core/management/commands/shell.py", line 69, in handle
    self.run_shell(shell=options['interface'])
  File "/home/vagrant/.virtualenvs/pulp3/lib/python3.5/site-packages/Django-1.8.14-py3.5.egg/django/core/management/commands/shell.py", line 61, in run_shell
    raise ImportError
ImportError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: Unicode-objects must be encoded before hashing
>>> 
```

If the value is not already raw bytes, it needs to be encoded.

Something like:
```
h = hashlib.sha256()
for name in sorted(f.name for f in self.natural_key_fields):
    h.update(name.encode('utf-8'))  # assuming name is also unicode
    value = getattr(self, name)
    # I honestly don't know the best way to be Python 2 and Python 3 compatible
    # Maybe this works? Or maybe this doesn't work at all and is stupid.
    if isinstance(value, bytes):
        h.update(value)
    else:
        h.update(value.encode('utf-8'))
    return h.hexdigest()
```